### PR TITLE
doc(ts): Add typescript documentation

### PR DESCRIPTION
**Pull request by Vorobeyko**
## Motivation/Description of the PR
- Description of this PR, which problem it solves
- Resolves #issueId (if applicable).

Applicable helpers:

- [ ] WebDriver
- [ ] Puppeteer
- [ ] Nightmare
- [ ] REST
- [ ] FileHelper
- [ ] Appium
- [ ] Protractor
- [ ] TestCafe
- [ ] Playwright

Applicable plugins:

- [ ] allure
- [ ] autoDelay
- [ ] autoLogin
- [ ] customLocator
- [ ] pauseOnFail
- [ ] puppeteerCoverage
- [ ] retryFailedStep
- [ ] screenshotOnFail
- [ ] selenoid
- [ ] stepByStepReport
- [ ] wdio

## Type of change

- [ ] :fire: Breaking changes
- [ ] :rocket: New functionality
- [ ] :bug: Bug fix
- [ ] :clipboard: Documentation changes/updates
- [ ] :hotsprings: Hot fix
- [ ] :hammer: Markdown files fix - not related to source code
- [ ] :nail_care: Polish code

## Checklist:

- [ ] Tests have been added
- [ ] Documentation has been added (Run `npm run docs`)
- [ ] Lint checking (Run `npm run lint`)
- [ ] Local tests are passed (Run `npm test`)

### CODE REVIEW

File: docs/typescript.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
+---
+permalink: /typescript
+title: TypeScript
+---
+
+# Typescript
+
+CodeceptJS supports [type declaration](https://github.com/codecept-js/CodeceptJS/tree/master/typings) for [TypeScript](https://www.typescriptlang.org/). It means that allow you to write your tests in TypeScript. Also, all of your custom steps could be write in TypeScript.
```
Comment (PeterNgTr) [inline docs/typescript.md:8]: It means now you can write your tests in TS. Also, all of your custom steps could be written in TS

File: docs/typescript.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+# Why TypeScript?
+
+With the TypeScript writing CodeceptJS tests becomes much easier. If correctly configured TypeScript in your project and correctly configured your IDE, then you can get the following features as result:
```
Comment (PeterNgTr) [inline docs/typescript.md:12]: If you configure TS properly in your project as well as your IDE, you will get the following features:

File: docs/typescript.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+- [Autocomplete (with InteleSence)](https://code.visualstudio.com/docs/editor/intellisense) - a tool that tries complete your thoughts for you and suggest you which function or property exist in class, what arguments can be passed to the method, what method return etc.
```
Comment (PeterNgTr) [inline docs/typescript.md:13]: I’m not sure InteleSence here is a tool or something? 😅

Comment (PeterNgTr) [inline docs/typescript.md:13]: A tool that streamlines your work by suggesting when you typing what function or property which exists in a class, what arguments can be passed to that method, what it returns, etc.

File: docs/typescript.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    "types": ["codeceptjs"],
+  },
+}
+```
+
+> You can find an example project with TypeScript and CodeceptJS on our project [typescript-boilerplate](https://github.com/codecept-js/typescript-boilerplate).
+
+### Set up steps.d.ts
+
+Not enough to configure `tsconfig.json` and `codecept.cong.js`. Also need to configure `steps.d.ts` file with custom steps. For it need to run def command:
```
Comment (PeterNgTr) [inline docs/typescript.md:78]: Configuring the tsconfig.json and codecept.conf.js is not enough, you will need to configure the step.d.ts file for custom steps. Just simply do this by running this command:

File: docs/typescript.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
+---
+permalink: /typescript
+title: TypeScript
+---
+
+# Typescript
+
+CodeceptJS supports [type declaration](https://github.com/codecept-js/CodeceptJS/tree/master/typings) for [TypeScript](https://www.typescriptlang.org/). It means now you can write your tests in TS. Also, all of your custom steps could be written in TS
```
Comment (DavertMik) [inline docs/typescript.md:8]:
```suggestion
CodeceptJS supports [type declaration](https://github.com/codecept-js/CodeceptJS/tree/master/typings) for [TypeScript](https://www.typescriptlang.org/). It means that you can write your tests in TS. Also, all of your custom steps can be written in TS
```

File: docs/typescript.md [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+# Why TypeScript?
+
+With the TypeScript writing CodeceptJS tests becomes much easier. If correctly configured TypeScript in your project and correctly configured your IDE, then you can get the following features as result:
+- [Autocomplete (with InteleSence)](https://code.visualstudio.com/docs/editor/intellisense) - a tool that tries complete your thoughts for you and suggest you which function or property exist in class, what arguments can be passed to the method, what method return etc. 
+Example:
+
+![Auto Complete](/img/Auto_comlete.gif)
+
+- Quick info about members - show info about each of your steps right in your tests. Example:
```
[the remark this answers is not in the document]
Comment (DavertMik) [reply docs/typescript.md:18]: probably it was items of a list

File: docs/typescript.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+![Quick Info](/img/Quick_info.gif)
+
+- Checks types - thanks to TypeScript support in CodeceptJS now allow to tests your tests. TypeScript can prevent some errors: 
+  - invalid type of variables passed to function;
+  - calls no-exist method from PageObject or `I` object;
+  - incorrect used CodeceptJS features; 
+
+
+## Getting Started
+### Instal TypeScipt
```
Comment (DavertMik) [inline docs/typescript.md:29]:
```suggestion
### Install TypeScipt
```

File: docs/typescript.md [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    "module": "commonjs",
+    "strictNullChecks": true,
+    "types": ["codeceptjs"],
+  },
+}
+```
+
+> You can find an example project with TypeScript and CodeceptJS on our project [typescript-boilerplate](https://github.com/codecept-js/typescript-boilerplate).
+
+### Set up steps.d.ts
```
Comment (DavertMik) [inline docs/typescript.md:76]:
```suggestion
### Set Up steps.d.ts
```

File: docs/typescript.md [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+  namespace Translation {
+    interface Actions {}
+  }
+}
+
+```
+
+## Types for custom helper or page object
+
+If you want to gets types for your [custom helper](https://codecept.io/helpers/#configuration), you can add their automatically with CodeceptJS command `npx codeceptjs def`.
```
Comment (DavertMik) [inline docs/typescript.md:100]:
```suggestion
If you want to get types for your [custom helper](https://codecept.io/helpers/#configuration), you can add their automatically with CodeceptJS command `npx codeceptjs def`. 
```

File: docs/typescript.md [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+  printMessage(msg: string) {
+    console.log(msg)
+  }
+}
+
+export = CustomHelper
+```
+
+Then you need to add this helper to your `codecept.conf.js` like in this [docs](https://codecept.io/helpers/#configuration).
+And then run to command `npx codeceptjs def`.
```
Comment (DavertMik) [inline docs/typescript.md:115]:
```suggestion
And then run the command `npx codeceptjs def`. 
```

File: docs/typescript.md [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+  interface I extends WithTranslation<Methods> {}
+  namespace Translation {
+    interface Actions {}
+  }
+}
+```
+
+And now you can use autocomplete on your test.
+
+Generation types for PageObject looks like for custom helper, but `steps.d.ts` will be looks like:
```
Comment (DavertMik) [inline docs/typescript.md:134]:
```suggestion
Generation types for PageObject looks like for a custom helper, but `steps.d.ts` will look like:
```

File: docs/typescript.md [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+title: TypeScript
+---
+
+# Typescript
+
+CodeceptJS supports [type declaration](https://github.com/codecept-js/CodeceptJS/tree/master/typings) for [TypeScript](https://www.typescriptlang.org/). It means that you can write your tests in TS. Also, all of your custom steps can be written in TS
+
+# Why TypeScript?
+
+With the TypeScript writing CodeceptJS tests becomes much easier. If you configure TS properly in your project as well as your IDE, you will get the following features:
```
Comment (DavertMik) [inline docs/typescript.md:12]: don't use word "easy" in documentation

File: docs/typescript.md [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+- [Autocomplete (with InteleSence)](https://code.visualstudio.com/docs/editor/intellisense) - a tool that streamlines your work by suggesting when you typing what function or property which exists in a class, what arguments can be passed to that method, what it returns, etc.
+Example:
+
+![Auto Complete](/img/Auto_comlete.gif)
+
+- Quick info about members - show info about each of your steps right in your tests. Example:
```
Comment (DavertMik) [inline docs/typescript.md:18]:
```suggestion
- To show additional information for a step in a test. Example:
```

File: docs/typescript.md [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+- Checks types - thanks to TypeScript support in CodeceptJS now allow to tests your tests. TypeScript can prevent some errors: 
+  - invalid type of variables passed to function;
+  - calls no-exist method from PageObject or `I` object;
+  - incorrect used CodeceptJS features; 
+
+
+## Getting Started
+### Install TypeScipt
+
+For writing tests in TypeScript you`ll need to install TypeScript and ts-node in your project.
```
Comment (DavertMik) [inline docs/typescript.md:31]:
```suggestion
For writing tests in TypeScript you`ll need to install `typescript` and `ts-node` into your project.
```

File: docs/typescript.md [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+  interface Methods extends Puppeteer {}
+  interface I extends WithTranslation<Methods> {}
+  namespace Translation {
+    interface Actions {}
+  }
+}
+
+```
+
+## Types for custom helper or page object
```
Comment (DavertMik) [inline docs/typescript.md:98]: please have separate chapters for helper & for page object

File: docs/typescript.md [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+---
+
+# Typescript
+
+CodeceptJS supports [type declaration](https://github.com/codecept-js/CodeceptJS/tree/master/typings) for [TypeScript](https://www.typescriptlang.org/). It means that you can write your tests in TS. Also, all of your custom steps can be written in TS
+
+# Why TypeScript?
+
+With the TypeScript writing CodeceptJS tests becomes much easier. If you configure TS properly in your project as well as your IDE, you will get the following features:
+- [Autocomplete (with InteleSence)](https://code.visualstudio.com/docs/editor/intellisense) - a tool that streamlines your work by suggesting when you typing what function or property which exists in a class, what arguments can be passed to that method, what it returns, etc.
```
Comment (pablopaul) [inline docs/typescript.md:13]: @Vorobeyko "IntelliSense" instead of "InteleSence"

File: docs/typescript.md [hunk 17]
```diff
@@ -0,0 +1,149 @@
+---
+permalink: /typescript
+title: TypeScript
+---
+
+# Typescript
+
+CodeceptJS supports [type declaration](https://github.com/codecept-js/CodeceptJS/tree/master/typings) for [TypeScript](https://www.typescriptlang.org/). It means that you can write your tests in TS. Also, all of your custom steps can be written in TS
```
Comment (pablopaul) [inline docs/typescript.md:8]:
```suggestion
CodeceptJS supports [type declaration](https://github.com/codecept-js/CodeceptJS/tree/master/typings) for [TypeScript](https://www.typescriptlang.org/). It means that you can write your tests in TS. Also, all of your custom steps can be written in TS.
```

File: docs/typescript.md (hunk 16 above)
Comment (pablopaul) [inline docs/typescript.md:13]:
```suggestion
- [Autocomplete (with IntelliSense)](https://code.visualstudio.com/docs/editor/intellisense) - a tool that streamlines your work by suggesting when you typing what function or property which exists in a class, what arguments can be passed to that method, what it returns, etc.
```

File: docs/typescript.md (hunk 12 above)
Comment (pablopaul) [inline docs/typescript.md:12]:
```suggestion
If you configure TS properly in your project as well as your IDE, you will get the following features:
```

File: docs/typescript.md [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+
+- To show additional information for a step in a test. Example:
+
+![Quick Info](/img/Quick_info.gif)
+
+- Checks types - thanks to TypeScript support in CodeceptJS now allow to tests your tests. TypeScript can prevent some errors: 
+  - invalid type of variables passed to function;
+  - calls no-exist method from PageObject or `I` object;
+  - incorrect used CodeceptJS features; 
+
```
Comment (pablopaul) [inline docs/typescript.md:26]:
```suggestion
```

File: docs/typescript.md [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+
+## Getting Started
+### Install TypeScipt
+
+For writing tests in TypeScript you`ll need to install `typescript` and `ts-node` into your project.
+
+`npm install typescript ts-node`
```
Comment (pablopaul) [inline docs/typescript.md:33]:
```suggestion
`npm install typescript ts-node -S`
```

File: docs/typescript.md [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
+
+### Configure codecept.conf.js
+
+To configure TypeScript in your project, you need to add [`ts-node/register`](https://github.com/TypeStrong/ts-node) on first line in your config. Like in the following config file:
```
Comment (pablopaul) [inline docs/typescript.md:37]:
```suggestion
To configure TypeScript in your project, you need to require [`ts-node/register`](https://github.com/TypeStrong/ts-node) on first line in your config. Like in the following config file:
```

File: docs/typescript.md [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
+    "types": ["codeceptjs"],
+  },
+}
+```
+
+> You can find an example project with TypeScript and CodeceptJS on our project [typescript-boilerplate](https://github.com/codecept-js/typescript-boilerplate).
+
+### Set Up steps.d.ts
+
+Configuring the `tsconfig.json` and `codecept.conf.js` is not enough, you will need to configure the `steps.d.ts` file for custom steps. Just simply do this by running this command::
```
Comment (pablopaul) [inline docs/typescript.md:78]:
```suggestion
Configuring the `tsconfig.json` and `codecept.conf.js` is not enough, you will need to configure the `steps.d.ts` file for custom steps. Just simply do this by running this command:
```

File: docs/typescript.md [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+  interface SupportObject { I: I }
+  interface Methods extends Puppeteer, CustomHelper {}
+  interface I extends WithTranslation<Methods> {}
+  namespace Translation {
+    interface Actions {}
+  }
+}
+```
+
+And now you can use autocomplete on your test.
```
Comment (pablopaul) [inline docs/typescript.md:132]:
```suggestion
```

File: docs/typescript.md [hunk 23]
```diff
@@ -0,0 +1,149 @@
...
+
+Generation types for PageObject looks like for a custom helper, but `steps.d.ts` will look like:
```
Comment (pablopaul) [inline docs/typescript.md:134]: @Vorobeyko what do you mean by "Generation types"?

Comment (Vorobeyko) [reply docs/typescript.md:134]: Maybe - "To generate types", sorry

Comment (pablopaul) [inline docs/typescript.md:134]:
```suggestion
Using `npx codeceptjs def` will also generate types for PageObjects. `steps.d.ts` will look like:
```

**Pull request merged**